### ISSUE TITLE
Keyboard select should not copy

### ISSUE THREAD

**Issue by sveniu** [Keruspe/GPaste#449]
When using the keyboard and either pressing Ctrl+A to select all or Shift+arrows/home/end, GPaste picks up the selection and puts it in the clipboard. If this was done with a mouse, then it is indeed expected that it's a primary selection and it should be picked up by the clipboard. But when done with a keyboard, it should not trigger a primary selection.

Notable exceptions:
* Firefox input fields do not copy on Ctrl+A; they do copy on Shift+arrows/home/end.
* Firefox text areas _do_ copy on Ctrl+A.
* Firefox text areas do not copy on Shift+pgup/pgdown or Ctrl+Shift+home/end. (Text Editor _does_ copy on Shift+pgup/pgdown or Ctrl+Shift+home/end.)

Steps to reproduce:
1. Start the GPaste 45 GUI
2. Start the Gnome Text Editor
3. Write a few characters or lines in the editor
4. Click Ctrl+A or Shift+arrows/home/end to select text

Expected behaviour: GPaste does not pick up the selection.

Actual behaviour: GPaste picks up the selection and puts it in the clipboard.

Impact: The classic "Ctrl+A to select all and then paste to replace" doesn't work. Instead of replacing the text, Ctrl+A copies the text you want to replace, and you just end up pasting that text again and making no changes.

I'm aware of the "Detect growing lines" option, but it does not seem to be relevant to this issue.

Environment:
* Ubuntu 24.04.1 with default Gnome 46 on Wayland 1.22.
* GPaste 45.
* Options:
  * Track clipboard changes: Y
  * Close UI on select: Y
  * Open UI window centered: N
  * Enable the gnome-shell extension: Y
  * Sync the daemon state with the extension's one: N
  * Primary selection affects history: Y
  * Synchronize clipboard with primary selection: Y
  * Images support: N
  * Trim items: N
  * Detect growing lines: N
  * Save history: Y

**Comment by Keruspe** [comment]
Hello,

That's what "primary selection affects history" is about. When you select something, it gets added to the history. Note that this isn't enabled by default

**Comment by Keruspe** [comment]
Note that GPaste isn't the one deciding whether something affects primary selection or not. It only reacts to system events. The application in which you select things is the one emitting the event.
GPaste only receives a notification telling it that the owner of the primary selection has changed, but no context as to how it happened (keyboard vs mouse vs whatever else)

**Comment by sveniu** [comment]
Hello! I have now disabled "Primary selection affects history", and the behaviour seems to remain the same: Ctrl+A in Text Editor copies to the clipboard.

As for your follow-up: That makes sense, and I'd be interested in figuring out what it is that makes Ctrl+A affect the primary selection "upstream", whether in Gnome or Wayland or something else. I'll try to understand what GPaste receives from, and then follow from there. If you have some hints on where to go, I'd greatly appreciate it.

**Comment by Keruspe** [comment]
Ah yes, that probably would be the “synchronize clipboard with primary
selection” which updates the clipboard when we get a notification about the
primary selection changing

**Comment by sveniu** [comment]
I've spent hours trying to replicate how GPaste picks up events, but I'm getting no activity at all.

I've tried both the attached Python snippet, and also this equivalent C example: https://github.com/bstpierre/gtk-examples/blob/master/c/clipboard_watch.c

How does GPaste actually get events? Is it not via "owner-change" in g_paste_clipboards_manager_activate? Does it have multiple strategies? Is the gnome-shell extension involved here?

[gtk_clipboard_test.txt](https://github.com/user-attachments/files/16976293/gtk_clipboard_test.txt)

**Comment by Keruspe** [comment]
It is through this event. It only works on X though, so we force the X backend at startup. It works on wayland thanks to mutter/gnome-shell acting as a bridge between wayland and Xwayland

**Comment by sveniu** [comment]
Thanks for that hint about gdk_set_allowed_backends! I've made it work by setting env var GDK_BACKEND=x11 on both the Python and C examples. (I tried using the Python equivalent `Gdk.set_allowed_backends("x11")` without effect, but that's probably just a local issue in my code.)

Now that I can reproduce the "unwanted" clipboard updates, I'll continue to follow the trace somewhere in Gnome.

Thanks again.
